File: src/cli_agent_orchestrator/services/macro_notation.py
"""The macro notation: one pinned grammar, and this is the authority.

Native-TUI-console §5.3.  Two parsers exist — this Python authority
(server-side, behind ``POST /macros/parse-notation``) and Lane B's
TypeScript live preview — and they are tested against the same golden
vectors (``test/fixtures/notation_vectors.json``), mirroring the digest
golden-vector precedent: the two sides cannot drift into spelling the
same macro two ways.

The grammar, exactly:

.. code-block:: text

    sequence := event (WS+ event)*
    event    := text | named | chord | repeat
    text     := '"' JSON-string '"'        # JSON escaping exactly; , + / \\ literal inside quotes
    named    := [a-z][a-z0-9-]*            # enter escape up down left right home end
                                           # page-up page-down delete insert tab backspace
    chord    := 'ctrl+' [a-z]              # ctrl+c ctrl+s … (D7 mapping)
    repeat   := (named|chord) '*' [1-9][0-9]*   # up*3; expansion counts toward the 32-event cap

Notation never touches disk (§5.1): the stored and transmitted
correctness boundary is the v3 event array, and this module's only
product is that array plus its readable preview.  Parse errors carry a
0-based character offset and a message; an unparseable or
unrepresentable macro cannot be saved or sent.
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from typing import Any, Dict, List, Optional

from cli_agent_orchestrator.services.control_input_contract import (
    MAX_SEQUENCE_EVENTS,
    MAX_SEQUENCE_TEXT_BYTES,
)

# Notation names to wire key names (§5.3).  The set is the §3.2 key set
# spelled in notation form; a name outside it is a parse error, never a
# guess (``BTab``, modified arrows, and F-keys have no notation spelling
# because the wire refuses them).
NOTATION_KEY_NAMES = {
    "enter": "Enter",
    "escape": "Escape",
    "up": "Up",
    "down": "Down",
    "left": "Left",
    "right": "Right",
    "home": "Home",
    "end": "End",
    "page-up": "PageUp",
    "page-down": "PageDown",
    "delete": "Delete",
    "insert": "Insert",
    "tab": "Tab",
    "backspace": "Backspace",
}

_NAMED_PATTERN = re.compile(r"[a-z][a-z0-9-]*")
_CHORD_PREFIX = "ctrl+"
_REPEAT_PATTERN = re.compile(r"[1-9][0-9]*")

# Modifier words a combination may be built from.  A combination of them
# (``ctrl+shift+x``, ``alt+x``, …) has no standard-mode byte encoding —
# tmux would inject the base key or a wrong encoding (§3.3) — so it is
# named and refused rather than misread as a single-modifier chord.
_MODIFIER_WORDS = frozenset({"ctrl", "alt", "meta", "cmd", "shift", "super"})
_COMBINATION_PATTERN = re.compile(r"[a-z0-9][a-z0-9+\-]*")


def _multi_modifier_failure(notation: str, pos: int) -> NotationResult:
    """The §3.3 refusal for a combination the byte stream cannot carry."""
    match = _COMBINATION_PATTERN.match(notation, pos)
    combination = match.group(0) if match is not None else notation[pos:]
    return _fail(
        pos,
        f"multi-modifier combination {combination!r} cannot be represented: no "
        "standard-mode terminal byte encoding exists for it (tmux would inject the "
        "base key or a wrong encoding), so it is refused, never approximated",
    )


# Text bytes a pane can never be sent honestly through the control path
# (the deployed literal screen): ESC and its C1 spelling synthesise
# escape sequences; CR/LF submit at a point the caller did not choose.
# The send path screens them again; the parse authority refuses them here
# so an unrepresentable macro cannot be saved.
_ILLEGAL_TEXT_CHARS = ("\x1b", "\x9b", "\r", "\n")

# A lone surrogate (a high surrogate not followed by a low one, or a low
# one not preceded by a high one) is valid JSON but not UTF-8-encodable.
_LONE_SURROGATE_RE = re.compile(
    r"[\ud800-\udbff](?![\udc00-\udfff])|(?<![\ud800-\udbff])[\udc00-\udfff]"
)


@dataclass(frozen=True)
class NotationError:
    """One parse failure: where, and what is wrong there."""

    offset: int
    message: str


@dataclass(frozen=True)
class NotationResult:
    """The parse answer: the events and preview, or the errors.

    ``errors`` is empty exactly when the parse succeeded; ``events`` and
    ``preview`` are then both set.  At most one error is reported — the
    parse fails fast at the first malformed token, because a parser that
    continues past an error would be guessing at intent from then on.
    """

    events: Optional[List[Dict[str, Any]]]
    preview: Optional[str]
    errors: List[NotationError]


def _fail(offset: int, message: str) -> NotationResult:
    return NotationResult(events=None, preview=None, errors=[NotationError(offset, message)])


def _chord_event(letter: str) -> Dict[str, Any]:
    """The wire event for one ``ctrl+<letter>`` token (D7).

    ``ctrl+c`` maps to the provider-agnostic ``key`` event ``C-c``; every
    other letter maps to a ``chord`` event (which the server admits only
    when the provider+build pins it — a send-time, per-terminal fact this
    parser cannot know and therefore does not check).
    """
    if letter == "c":
        return {"type": "key", "key": "C-c"}
    return {"type": "chord", "chord": f"C-{letter}"}


def parse_notation(notation: str) -> NotationResult:
    """Parse one notation string into its v3 events and readable preview.

    The single authority for the grammar above.  Never approximates: a
    token that is not exactly one of the four event forms is an error
    with an offset, and an unrepresentable one (an over-cap expansion, an
    over-budget text, a control character the pane cannot be sent) is
    refused rather than approximated into something else.
    """
    if not isinstance(notation, str):
        return _fail(0, f"notation must be a string, got {type(notation).__name__}")
    length = len(notation)
    pos = 0
    while pos < length and notation[pos].isspace():
        pos += 1
    if pos == length:
        return _fail(pos, "a macro names at least one event")

    events: List[Dict[str, Any]] = []
    text_bytes = 0

    while pos < length:
        event_pos = pos
        base: Optional[Dict[str, Any]] = None
        repeatable = False
        char = notation[pos]

        if char == '"':
            # A JSON string, scanned to its closing quote (backslash
            # escapes honoured) and decoded by JSON's own rules, exactly.
            index = pos + 1
            closed = False
            while index < length:
                c = notation[index]
                if c == "\\":
                    index += 2
                    continue
                if c == '"':
                    closed = True
                    break
                index += 1
            if not closed:
                return _fail(event_pos, "unterminated string: a text token is a JSON string")
            token = notation[event_pos : index + 1]
            try:
                decoded = json.loads(token)
            except json.JSONDecodeError as exc:
                return _fail(
                    event_pos + exc.pos,
                    f"invalid JSON string: {exc.msg}; text uses JSON escaping exactly "
                    "(comma, plus, slash and backslash are literal inside quotes)",
                )
            if not isinstance(decoded, str) or decoded == "":
                return _fail(event_pos, "a text event must be a non-empty string")
            for illegal in _ILLEGAL_TEXT_CHARS:
                at = decoded.find(illegal)
                if at != -1:
                    return _fail(
                        event_pos,
                        "the text contains a control character (ESC, C1 CSI, CR, or LF) "
                        "the control path can never send honestly; an unrepresentable "
                        "macro is refused, never approximated",
                    )
            if _LONE_SURROGATE_RE.search(decoded) is not None:
                # A lone surrogate parses as valid JSON but is not
                # UTF-8-encodable, so it can never become a wire event's
                # text; screening here keeps the failure in the ordinary
                # offset-bearing shape (the endpoint answers 422, never
                # 500) — the same defect class as the r11 repeat guard.
                return _fail(
                    event_pos,
                    "the text contains a lone surrogate that is not UTF-8-encodable; "
                    "an unrepresentable macro is refused, never approximated",
                )
            text_bytes += len(decoded.encode("utf-8"))
            if text_bytes > MAX_SEQUENCE_TEXT_BYTES:
                return _fail(
                    event_pos,
                    f"this text pushes the sequence past the {MAX_SEQUENCE_TEXT_BYTES}-byte "
                    "aggregate text cap; a macro is a short control burst, not a document",
                )
            base = {"type": "text", "text": decoded}
            pos = index + 1
        elif notation.startswith(_CHORD_PREFIX, pos):
            letter_at = pos + len(_CHORD_PREFIX)
            if letter_at < length:
                word = _NAMED_PATTERN.match(notation, letter_at)
                if (
                    word is not None
                    and word.group(0) in _MODIFIER_WORDS
                    and (word.end() >= length or notation[word.end()] in "+ \t\n")
                ):
                    # ``ctrl+shift+x`` and friends: a modifier where the
                    # chord's single letter must be.
                    return _multi_modifier_failure(notation, pos)
            if letter_at >= length or not ("a" <= notation[letter_at] <= "z"):
                return _fail(
                    letter_at,
                    "a chord is 'ctrl+' followed by one letter a-z; multi-modifier and "
                    "non-letter chords are unrepresentable and refused, never approximated",
                )
            base = _chord_event(notation[letter_at])
            repeatable = True
            pos = letter_at + 1
        else:
            match = _NAMED_PATTERN.match(notation, pos)
            if match is None:
                return _fail(
                    pos,
                    f"expected a quoted text, a named key, or a ctrl+<letter> chord, "
                    f"not {char!r}; known key names are {sorted(NOTATION_KEY_NAMES)}",
                )
            name = match.group(0)
            if name in _MODIFIER_WORDS and match.end() < length and notation[match.end()] == "+":
                # ``alt+x``, ``meta+x`` and friends at event position.
                return _multi_modifier_failure(notation, pos)
            if name not in NOTATION_KEY_NAMES:
                return _fail(
                    pos,
                    f"unknown named key {name!r}; known key names are "
                    f"{sorted(NOTATION_KEY_NAMES)} — unlisted keys (BTab, modified "
                    "arrows, F-keys) are refused, never approximated",
                )
            base = {"type": "key", "key": NOTATION_KEY_NAMES[name]}
            repeatable = True
            pos = match.end()

        count = 1
        if repeatable and pos < length and notation[pos] == "*":
            repeat_pos = pos
            digits = _REPEAT_PATTERN.match(notation, pos + 1)
            if digits is None:
                return _fail(
                    repeat_pos + 1,
                    "a repeat count is a positive integer written [1-9][0-9]* "
                    "(zero and empty counts are malformed, not no-ops)",
                )
            count_text = digits.group(0)
            if len(count_text) > 2:
                # A count of 100+ can never fit the 32-event budget, even
                # in an empty sequence — fail BEFORE the integer
                # conversion (CPython refuses over-long digit strings with
                # a bare ValueError from its int-max-str-digits guard).
                # r11/Sol: the failure keeps the ordinary offset-bearing
                # shape, and the endpoint answers 422, never 500; the
                # message embeds no token, so it is bounded by
                # construction.
                return _fail(
                    event_pos,
                    f"this event brings the sequence past the {MAX_SEQUENCE_EVENTS}-event "
                    "cap; a repeat expansion counts every event it stands for",
                )
            count = int(count_text)
            pos = digits.end()

        if len(events) + count > MAX_SEQUENCE_EVENTS:
            return _fail(
                event_pos,
                f"this event brings the sequence past the {MAX_SEQUENCE_EVENTS}-event "
                "cap; a repeat expansion counts every event it stands for",
            )
        events.extend(base for _ in range(count))

        if pos < length:
            if not notation[pos].isspace():
                return _fail(pos, f"expected whitespace between events, not {notation[pos]!r}")
            while pos < length and notation[pos].isspace():
                pos += 1

    return NotationResult(events=events, preview=preview_sequence(events), errors=[])


def _preview_token(event: Dict[str, Any]) -> str:
    """One event's canonical preview token, extending the deployed recorder.

    Text renders as its JSON string (the notation's own spelling, so the
    preview round-trips); a chord renders ``[Ctrl+X]``; the provider-
    agnostic interrupt key renders ``[Ctrl+C]``; every other key renders
    its wire name in brackets.
    """
    if event["type"] == "text":
        return json.dumps(event["text"], ensure_ascii=False)
    if event["type"] == "chord":
        return f"[Ctrl+{event['chord'][2:].upper()}]"
    if event["key"] == "C-c":
        return "[Ctrl+C]"
    return f"[{event['key']}]"


def preview_sequence(events: List[Dict[str, Any]]) -> str:
    """The canonical one-line preview: ``"text" [Enter] [Up]×3 [Ctrl+S]``.

    A run of two or more identical key/chord events collapses to
    ``[Name]×N`` — the normalized form, whether the run came from a
    repeat token or from spelled-out repetition.  Text events are never
    collapsed: two adjacent text events are different content, not a
    repeat.
    """
    tokens: List[str] = []
    index = 0
    while index < len(events):
        event = events[index]
        run_end = index + 1
        if event["type"] != "text":
            while run_end < len(events) and events[run_end] == event:
                run_end += 1
        token = _preview_token(event)
        run = run_end - index
        tokens.append(f"{token}×{run}" if run > 1 else token)
        index = run_end
    return " ".join(tokens)

File: web/src/lib/macroNotation.ts
/**
 * Operator-macro notation: the §5.3 editing-surface grammar (TS live preview).
 *
 * The server (`services/macro_notation.py`, Lane A's canonical parser) is the
 * authority: it decides what may be saved or sent. This TypeScript preview is
 * its mirror — same tokenization, same messages, same offsets — pinned
 * byte-for-byte by the shared golden vectors in
 * `test/fixtures/notation_vectors.json`, mirroring the digest golden-vector
 * precedent. Any grammar change lands in both parsers and the vectors in one
 * change — a needed-but-absent grammar item is a spec amendment, never a
 * frontend invention.
 *
 * Grammar (pinned, §5.3):
 *
 *   sequence := event (WS+ event)*
 *   event    := text | named | chord | repeat
 *   text     := '"' JSON-string '"'      — JSON escaping exactly
 *   named    := [a-z][a-z0-9-]*          — the fourteen names in NAMED_KEYS
 *   chord    := 'ctrl+' [a-z]            — ctrl+c → key C-c; others → chord
 *   repeat   := (named|chord) '*' [1-9][0-9]*   — expansion counts toward
 *                                                 the 32-event cap
 *
 * Known residuals (documented, never hit by realistic editor input):
 * Python's str.isspace() and JS's \s differ on U+0085 and U+FEFF; malformed
 * \uXXXX escapes report Python's "Invalid \escape" rather than CPython's
 * \u-specific wording. The server authority is the gate for both.
 */

import type { SequenceEvent } from './sequenceRecorder'
import { MAX_SEQUENCE_EVENTS, MAX_SEQUENCE_TEXT_BYTES } from './sequenceRecorder'

/** The fourteen named keys of the §5.3 grammar, notation name → wire name. */
export const NAMED_KEYS: Record<string, string> = {
  enter: 'Enter',
  escape: 'Escape',
  up: 'Up',
  down: 'Down',
  left: 'Left',
  right: 'Right',
  home: 'Home',
  end: 'End',
  'page-up': 'PageUp',
  'page-down': 'PageDown',
  delete: 'Delete',
  insert: 'Insert',
  tab: 'Tab',
  backspace: 'Backspace',
}

/** Wire name → notation name, for the canonical renderer. */
const WIRE_TO_NOTATION: Record<string, string> = Object.fromEntries(
  Object.entries(NAMED_KEYS).map(([name, wire]) => [wire, name]),
)

// The sorted known-names list rendered exactly as Python's str(sorted(...)).
const KNOWN_NAMES =
  "['backspace', 'delete', 'down', 'end', 'enter', 'escape', 'home', 'insert', " +
  "'left', 'page-down', 'page-up', 'right', 'tab', 'up']"

// Modifier words a combination may be built from (§3.3): a combination of
// them has no standard-mode byte encoding, so it is named and refused.
const MODIFIER_WORDS = new Set(['ctrl', 'alt', 'meta', 'cmd', 'shift', 'super'])

const NAMED_RE = /[a-z][a-z0-9-]*/y
const COMBINATION_RE = /[a-z0-9][a-z0-9+\-]*/y
const REPEAT_RE = /[1-9][0-9]*/y

// Chars that end the modifier check after a modifier word (Python's
// ``notation[word.end()] in "+ \t\n"``).
const MODIFIER_ENDERS = new Set(['+', ' ', '\t', '\n'])

// Text bytes the control path can never send honestly (ESC, C1 CSI, CR, LF).
const ILLEGAL_TEXT_CHARS = ['\x1b', '\x9b', '\r', '\n']

const LONE_SURROGATE_RE = /[\uD800-\uDBFF](?![\uDC00-\uDFFF])|(?<![\uD800-\uDBFF])[\uDC00-\uDFFF]/

const CAP_MESSAGE =
  `this event brings the sequence past the ${MAX_SEQUENCE_EVENTS}-event cap; ` +
  'a repeat expansion counts every event it stands for'

export interface NotationErrorInfo {
  offset: number
  message: string
}

/** One parse/render failure carrying the §5.3 (offset, message) pair. */
export class NotationParseError extends Error {
  readonly offset: number

  constructor(offset: number, message: string) {
    super(message)
    this.name = 'NotationParseError'
    this.offset = offset
  }

  asInfo(): NotationErrorInfo {
    return { offset: this.offset, message: this.message }
  }
}

const encoder = new TextEncoder()

function isSpace(ch: string): boolean {
  return /\s/.test(ch)
}

function multiModifierFailure(notation: string, pos: number): NotationParseError {
  COMBINATION_RE.lastIndex = pos
  const match = COMBINATION_RE.exec(notation)
  const combination = match !== null ? match[0] : notation.slice(pos)
  return new NotationParseError(
    pos,
    `multi-modifier combination '${combination}' cannot be represented: no ` +
      'standard-mode terminal byte encoding exists for it (tmux would inject the ' +
      'base key or a wrong encoding), so it is refused, never approximated',
  )
}

/**
 * Scan and decode the JSON string opening at `start`, reproducing the Python
 * authority's error taxonomy (unterminated / invalid escape / invalid
 * control character) with Python-identical offsets and messages for the
 * pinned cases.
 */
function scanJsonString(notation: string, start: number): [string, number] {
  const n = notation.length
  let k = start + 1
  let closed = false
  while (k < n) {
    const ch = notation[k]
    if (ch === '\\') {
      k += 2
      continue
    }
    if (ch === '"') {
      closed = true
      break
    }
    k += 1
  }
  if (!closed) {
    throw new NotationParseError(start, 'unterminated string: a text token is a JSON string')
  }
  const token = notation.slice(start, k + 1)
  // Validate escape sequences and raw control characters with CPython's
  // messages (the authority's ``json`` errors, offset = the offending index
  // inside the token, reported at event_pos + index).
  for (let i = 1; i < token.length - 1; i += 1) {
    const ch = token[i]
    if (ch === '\\') {
      const next = token[i + 1]
      if (next === undefined || !'"\\/bfnrtu'.includes(next)) {
        throw new NotationParseError(
          start + i,
          'invalid JSON string: Invalid \\escape; text uses JSON escaping exactly ' +
            '(comma, plus, slash and backslash are literal inside quotes)',
        )
      }
      i += 1
      continue
    }
    if (ch < ' ') {
      throw new NotationParseError(
        start + i,
        'invalid JSON string: Invalid control character at; text uses JSON escaping ' +
          'exactly (comma, plus, slash and backslash are literal inside quotes)',
      )
    }
  }
  const decoded = JSON.parse(token) as string
  return [decoded, k + 1]
}

/**
 * Parse §5.3 notation into a v3 event array. Throws NotationParseError with
 * an offset and message on any failure; at most one error is reported — the
 * parse fails fast at the first malformed token.
 */
export function parseNotation(notation: string): SequenceEvent[] {
  const length = notation.length
  let pos = 0
  while (pos < length && isSpace(notation[pos])) pos += 1
  if (pos === length) {
    throw new NotationParseError(pos, 'a macro names at least one event')
  }

  const events: SequenceEvent[] = []
  let textBytes = 0

  while (pos < length) {
    const eventPos = pos
    let event: SequenceEvent | null = null
    let repeatable = false
    const char = notation[pos]

    if (char === '"') {
      const [decoded, end] = scanJsonString(notation, pos)
      if (decoded === '') {
        throw new NotationParseError(eventPos, 'a text event must be a non-empty string')
      }
      for (const illegal of ILLEGAL_TEXT_CHARS) {
        if (decoded.includes(illegal)) {
          throw new NotationParseError(
            eventPos,
            'the text contains a control character (ESC, C1 CSI, CR, or LF) ' +
              'the control path can never send honestly; an unrepresentable ' +
              'macro is refused, never approximated',
          )
        }
      }
      if (LONE_SURROGATE_RE.test(decoded)) {
        throw new NotationParseError(
          eventPos,
          'the text contains a lone surrogate that is not UTF-8-encodable; ' +
            'an unrepresentable macro is refused, never approximated',
        )
      }
      textBytes += encoder.encode(decoded).length
      if (textBytes > MAX_SEQUENCE_TEXT_BYTES) {
        throw new NotationParseError(
          eventPos,
          `this text pushes the sequence past the ${MAX_SEQUENCE_TEXT_BYTES}-byte ` +
            'aggregate text cap; a macro is a short control burst, not a document',
        )
      }
      event = { type: 'text', text: decoded }
      pos = end
    } else if (notation.startsWith('ctrl+', pos)) {
      const letterAt = pos + 'ctrl+'.length
      if (letterAt < length) {
        NAMED_RE.lastIndex = letterAt
        const word = NAMED_RE.exec(notation)
        if (
          word !== null &&
          MODIFIER_WORDS.has(word[0]) &&
          (NAMED_RE.lastIndex >= length || MODIFIER_ENDERS.has(notation[NAMED_RE.lastIndex]))
        ) {
          // ``ctrl+shift+x`` and friends: a modifier where the chord's
          // single letter must be.
          throw multiModifierFailure(notation, pos)
        }
      }
      if (letterAt >= length || !(notation[letterAt] >= 'a' && notation[letterAt] <= 'z')) {
        throw new NotationParseError(
          letterAt,
          "a chord is 'ctrl+' followed by one letter a-z; multi-modifier and " +
            'non-letter chords are unrepresentable and refused, never approximated',
        )
      }
      const letter = notation[letterAt]
      event = letter === 'c' ? { type: 'key', key: 'C-c' } : { type: 'chord', chord: `C-${letter}` }
      repeatable = true
      pos = letterAt + 1
    } else {
      NAMED_RE.lastIndex = pos
      const match = NAMED_RE.exec(notation)
      if (match === null) {
        throw new NotationParseError(
          pos,
          `expected a quoted text, a named key, or a ctrl+<letter> chord, ` +
            `not '${char}'; known key names are ${KNOWN_NAMES}`,
        )
      }
      const name = match[0]
      if (MODIFIER_WORDS.has(name) && NAMED_RE.lastIndex < length && notation[NAMED_RE.lastIndex] === '+') {
        // ``alt+x``, ``meta+x`` and friends at event position.
        throw multiModifierFailure(notation, pos)
      }
      const wire = NAMED_KEYS[name]
      if (wire === undefined) {
        throw new NotationParseError(
          pos,
          `unknown named key '${name}'; known key names are ${KNOWN_NAMES} — ` +
            'unlisted keys (BTab, modified arrows, F-keys) are refused, never approximated',
        )
      }
      event = { type: 'key', key: wire }
      repeatable = true
      pos = NAMED_RE.lastIndex
    }

    let count = 1
    if (repeatable && pos < length && notation[pos] === '*') {
      const repeatPos = pos
      REPEAT_RE.lastIndex = pos + 1
      const digits = REPEAT_RE.exec(notation)
      if (digits === null) {
        throw new NotationParseError(
          repeatPos + 1,
          'a repeat count is a positive integer written [1-9][0-9]* ' +
            '(zero and empty counts are malformed, not no-ops)',
        )
      }
      if (digits[0].length > 2) {
        // A count of 100+ can never fit the 32-event budget, even in an
        // empty sequence — fail before the numeric conversion (huge digit
        // strings lose precision/overflow to Infinity). r11: the failure
        // keeps the ordinary offset-bearing shape; the message embeds no
        // token, so it is bounded by construction.
        throw new NotationParseError(eventPos, CAP_MESSAGE)
      }
      count = Number(digits[0])
      pos = REPEAT_RE.lastIndex
    }

    if (events.length + count > MAX_SEQUENCE_EVENTS) {
      throw new NotationParseError(eventPos, CAP_MESSAGE)
    }
    for (let k = 0; k < count; k += 1) events.push({ ...(event as SequenceEvent) })

    if (pos < length) {
      if (!isSpace(notation[pos])) {
        throw new NotationParseError(pos, `expected whitespace between events, not '${notation[pos]}'`)
      }
      while (pos < length && isSpace(notation[pos])) pos += 1
    }
  }

  return events
}

/** Non-throwing parse for the live editor: events, or the §5.3 errors. */
export function tryParseNotation(
  notation: string,
): { ok: true; events: SequenceEvent[] } | { ok: false; errors: NotationErrorInfo[] } {
  try {
    return { ok: true, events: parseNotation(notation) }
  } catch (error) {
    if (error instanceof NotationParseError) {
      return { ok: false, errors: [error.asInfo()] }
    }
    throw error
  }
}

function eventNotation(event: SequenceEvent): string {
  if (event.type === 'text') {
    // Canonical text form: JSON escaping exactly, non-ASCII literal.
    return JSON.stringify(event.text ?? '')
  }
  if (event.type === 'key') {
    if (event.key === 'C-c') return 'ctrl+c'
    const name = event.key !== undefined ? WIRE_TO_NOTATION[event.key] : undefined
    if (name === undefined) {
      throw new Error(`key ${JSON.stringify(event.key)} has no notation name`)
    }
    return name
  }
  if (event.type === 'chord') {
    const match = /^C-([a-z])$/.exec(event.chord ?? '')
    // ctrl+c parses to key C-c (D7), so a chord C-c has no faithful
    // notation form — rendering one would round-trip to a different event.
    if (match === null || match[1] === 'c') {
      throw new Error(`chord ${JSON.stringify(event.chord)} has no notation form`)
    }
    return `ctrl+${match[1]}`
  }
  throw new Error(`event type ${JSON.stringify(event.type)} has no notation form`)
}

function sameNonTextEvent(a: SequenceEvent, b: SequenceEvent): boolean {
  return (
    a.type !== 'text' &&
    b.type !== 'text' &&
    a.type === b.type &&
    a.key === b.key &&
    a.chord === b.chord
  )
}

/**
 * Render the canonical notation for a validated v3 event array (client-side:
 * the §7.4 recorder keeps its tokens and notation in sync through this and
 * the pinned parser; the server never renders notation). Runs of two or more
 * identical non-text events fold to `name*N`; text events never fold.
 */
export function renderNotation(events: SequenceEvent[]): string {
  const tokens: string[] = []
  let i = 0
  while (i < events.length) {
    const event = events[i]
    const token = eventNotation(event)
    if (event.type === 'text') {
      tokens.push(token)
      i += 1
      continue
    }
    let runEnd = i
    while (runEnd < events.length && sameNonTextEvent(events[runEnd], event)) runEnd += 1
    const run = runEnd - i
    tokens.push(run >= 2 ? `${token}*${run}` : token)
    i = runEnd
  }
  return tokens.join(' ')
}

/** One event's preview token: `"text"`, `[Enter]`, `[Ctrl+S]`. */
function previewTokenOf(event: SequenceEvent): string {
  if (event.type === 'text') return JSON.stringify(event.text ?? '')
  if (event.type === 'chord') {
    const chord = event.chord ?? ''
    const letter = chord.startsWith('C-') ? chord.slice(2) : chord
    return `[Ctrl+${letter.toUpperCase()}]`
  }
  if (event.key === 'C-c') return '[Ctrl+C]'
  return `[${event.key ?? ''}]`
}

/** The §5.3 normalized preview: `"text" [Enter] [Up]×3 [Ctrl+S]`. */
export function renderPreview(events: SequenceEvent[]): string {
  const tokens: string[] = []
  let i = 0
  while (i < events.length) {
    const event = events[i]
    const token = previewTokenOf(event)
    if (event.type === 'text') {
      tokens.push(token)
      i += 1
      continue
    }
    let runEnd = i
    while (runEnd < events.length && sameNonTextEvent(events[runEnd], event)) runEnd += 1
    const run = runEnd - i
    tokens.push(run >= 2 ? `${token}×${run}` : token)
    i = runEnd
  }
  return tokens.join(' ')
}
